src/lib/homestore_backend/hs_homeobject.cpp

```diff
@@ -134,17 +158,38 @@ void HSHomeObject::init_homestore() {
         RELEASE_ASSERT(!_our_id.is_nil(), "Received no SvcId and need FORMAT!");
         LOGW("We are starting for the first time on [{}], Formatting!!", to_string(_our_id));
 
-        HomeStore::instance()->format_and_start({
-            {HS_SERVICE::META, hs_format_params{.size_pct = 5.0}},
-            {HS_SERVICE::LOG, hs_format_params{.size_pct = 10.0, .chunk_size = 32 * Mi}},
-            {HS_SERVICE::REPLICATION,
-             hs_format_params{.size_pct = 79.0,
-                              .num_chunks = 65000,
-                              .block_size = _data_block_size,
-                              .alloc_type = blk_allocator_type_t::append,
-                              .chunk_sel_type = chunk_selector_type_t::CUSTOM}},
-            {HS_SERVICE::INDEX, hs_format_params{.size_pct = 5.0}},
-        });
+        if (has_data_dev && has_fast_dev) {
+            // Hybrid mode
+            LOGD("Has both Data and Fast, running with Hybrid mode");
+            HomeStore::instance()->format_and_start({
+                {HS_SERVICE::META, hs_format_params{.dev_type = HSDevType::Fast, .size_pct = 9.0}},
+                {HS_SERVICE::LOG,
+                 hs_format_params{.dev_type = HSDevType::Fast, .size_pct = 45.0, .chunk_size = 32 * Mi}},
+                {HS_SERVICE::INDEX, hs_format_params{.dev_type = HSDevType::Fast, .size_pct = 45.0}},
+                {HS_SERVICE::REPLICATION,
+                 hs_format_params{.dev_type = HSDevType::Data,
+                                  .size_pct = 99.0,
+                                  .num_chunks = 512,
```

Comment:
There are some nits in HS which will result a chunk smaller than 16MB as well as vdev larger than input size.

I am not sure if the io_uring error has anything to do with this but down the num_chunks to 512 solved the problem.

Will revisit after the https://github.com/eBay/HomeStore/issues/377 fixed

Reply:
```
[04/12/24 11:00:15-07:00] [I] [922219] [raft_repl_service.cpp:181:create_state_mgr] Creating RAFT state manager for server_id=47768266 group_id=12bcd86e-2cd2-4ee0-98af-0cda3b2799b8
[04/12/24 11:00:15-07:00] [I] [922249] [iomgr_timer.cpp:179:setup_timer_fd] Creating recurring per-thread timer fd 60 and adding it into fd poll list
[04/12/24 11:00:15-07:00] [I] [922219] [log_dev.cpp:656:create_new_log_store] Created log store log_dev=0 log_store=0
[04/12/24 11:00:16-07:00] [I] [922219] [replication_state_machine.cpp:35:on_pre_commit] on_pre_commit with lsn:2
[04/12/24 11:00:16-07:00] [I] [922252] [replication_state_machine.cpp:7:on_commit] applying raft log commit with lsn:2
[04/12/24 11:00:16-07:00] [I] [922252] [hs_pg_manager.cpp:138:on_create_pg_message_commit] Index table created for pg 1 uuid 448be9fb-d40f-4097-ae63-99275777891a
[04/12/24 11:00:16-07:00] [E] [922223] [uring_drive_interface.cpp:502:handle_completions] Error in completion of io, iocb=0x6110001880c0, result=-22, retry=0
[04/12/24 11:00:16-07:00] [E] [922223] [uring_drive_interface.cpp:502:handle_completions] Error in completion of io, iocb=0x6110001880c0, result=-22, retry=1
[04/12/24 11:00:16-07:00] [E] [922223] [uring_drive_interface.cpp:502:handle_completions] Error in completion of io, iocb=0x6110001880c0, result=-22, retry=2
[04/12/24 11:00:16-07:00] [E] [922223] [uring_drive_interface.cpp:502:handle_completions] Error in completion of io, iocb=0x6110001880c0, result=-22, retry=3
[04/12/24 11:00:16-07:00] [E] [922223] [uring_drive_interface.cpp:502:handle_completions] Error in completion of io, iocb=0x6110001880c0, result=-22, retry=4
homestore_test: /var/home/centos/.conan/data/iomgr/11.2.2-33/oss/master/build/8f6113204e39cf5069af1f6032b873fa6df4d478/src/lib/interfaces/uring_drive_interface.cpp:506: void iomgr::UringDriveInterface::handle_completions(): Assertion `0' failed.
[04/12/24 11:00:16-07:00] [C] [922223] Don't expect op=WRITE retry exceed limit=3
*** Aborted at 1712944816 (Unix time, try 'date -d @1712944816') ***
*** Signal 6 (SIGABRT) (0x3e8000e126b) received by PID 922219 (pthread TID 0x7f9c1dd5d600) (linux TID 922223) (maybe from PID 922219, UID 1000) (code: -6), stack trace: ***
    @ 0000000006b81580 _ZN5folly10symbolizer12_GLOBAL__N_118innerSignalHandlerEiP9siginfo_tPv
                       /var/home/centos/.conan/data/folly/nu2.2023.12.18.00/_/_/build/1f94d017cff1826b63c1a8dd2b63251bdcf2c15c/src/folly/experimental/symbolizer/SignalHandler.cpp:449
    @ 0000000006b81666 _ZN5folly10symbolizer12_GLOBAL__N_113signalHandlerEiP9siginfo_tPv
                       /var/home/centos/.conan/data/folly/nu2.2023.12.18.00/_/_/build/1f94d017cff1826b63c1a8dd2b63251bdcf2c15c/src/folly/experimental/symbolizer/SignalHandler.cpp:470
    @ 000000000004251f (unknown)
    @ 00000000000969fc pthread_kill
    @ 0000000000042475 raise
    @ 00000000000287f2 abort
    @ 000000000002871a (unknown)
    @ 0000000000039e95 __assert_fail
    @ 00000000066ddb4e _ZN5iomgr19UringDriveInterface18handle_completionsEv
                       /var/home/centos/.conan/data/iomgr/11.2.2-33/oss/master/build/8f6113204e39cf5069af1f6032b873fa6df4d478/src/lib/interfaces/uring_drive_interface.cpp:506
    @ 00000000066d82ac _ZZN5iomgr19uring_drive_channelC4EPNS_19UringDriveInterfaceEENKUlvE0_clEv
                       /var/home/centos/.conan/data/iomgr/11.2.2-33/oss/master/build/8f6113204e39cf5069af1f6032b873fa6df4d478/src/lib/interfaces/uring_drive_interface.cpp:56
    @ 00000000066e0c93 _ZSt13__invoke_implIvRZN5iomgr19uring_drive_channelC4EPNS0_19UringDriveInterfaceEEUlvE0_JEET_St14__invoke_otherOT0_DpOT1_
                       /usr/include/c++/11/bits/invoke.h:61
                       -> /var/home/centos/.conan/data/iomgr/11.2.2-33/oss/master/build/8f6113204e39cf5069af1f6032b873fa6df4d478/src/lib/interfaces/uring_drive_interface.cpp
    @ 00000000066e06c8 _ZSt10__invoke_rIvRZN5iomgr19uring_drive_channelC4EPNS0_19UringDriveInterfaceEEUlvE0_JEENSt9enable_ifIX16is_invocable_r_vIT_T0_DpT1_EES7_E4typeEOS8_DpOS9_
                       /usr/include/c++/11/bits/invoke.h:111
                       -> /var/home/centos/.conan/data/iomgr/11.2.2-33/oss/master/build/8f6113204e39cf5069af1f6032b873fa6df4d478/src/lib/interfaces/uring_drive_interface.cpp
    @ 00000000066dfaf8 _ZNSt17_Function_handlerIFvvEZN5iomgr19uring_drive_channelC4EPNS1_19UringDriveInterfaceEEUlvE0_E9_M_invokeERKSt9_Any_data
                       /usr/include/c++/11/bits/std_function.h:290
                       -> /var/home/centos/.conan/data/iomgr/11.2.2-33/oss/master/build/8f6113204e39cf5069af1f6032b873fa6df4d478/src/lib/interfaces/uring_drive_interface.cpp
    @ 0000000006167c3b _ZNKSt8functionIFvvEEclEv
                       /usr/include/c++/11/bits/std_function.h:590
                       -> /var/home/centos/.conan/data/homestore/6.2.1-47/oss/master/build/98a6c2ddffe4d1fb00eb495a8407dc88f8f61572/src/lib/homestore.cpp
    @ 0000000006707a68 _ZN5iomgr9IOReactor11listen_onceEv
                       /var/home/centos/.conan/data/iomgr/11.2.2-33/oss/master/build/8f6113204e39cf5069af1f6032b873fa6df4d478/src/lib/reactor/reactor.cpp:120
    @ 0000000006706f50 _ZN5iomgr9IOReactor3runEimjRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEERKSt8functionIFbRKSt10shared_ptrIKNS_8IODeviceEEEEOS9_IFvbEE
                       /var/home/centos/.conan/data/iomgr/11.2.2-33/oss/master/build/8f6113204e39cf5069af1f6032b873fa6df4d478/src/lib/reactor/reactor.cpp:71
    @ 0000000006614b3a _ZN5iomgr9IOManager12_run_io_loopEimjRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEERKSt8functionIFbRKSt10shared_ptrIKNS_8IODeviceEEEEOS9_IFvbEE
                       /var/home/centos/.conan/data/iomgr/11.2.2-33/oss/master/build/8f6113204e39cf5069af1f6032b873fa6df4d478/src/lib/iomgr.cpp:334
    @ 00000000066ee3e7 _ZZN5iomgr18IOManagerEpollImpl19create_reactor_implERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEmjiOSt8functionIFvbEEENUlvE_clEv
                       /var/home/centos/.conan/data/iomgr/11.2.2-33/oss/master/build/8f6113204e39cf5069af1f6032b873fa6df4d478/src/lib/epoll/iomgr_impl_epoll.cpp:17
    @ 00000000066ef1e5 _ZSt13__invoke_implIvZN5iomgr18IOManagerEpollImpl19create_reactor_implERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEmjiOSt8functionIFvbEEEUlvE_JEET_St14__invoke_otherOT0_DpOT1_
                       /usr/include/c++/11/bits/invoke.h:61
                       -> /var/home/centos/.conan/data/iomgr/11.2.2-33/oss/master/build/8f6113204e39cf5069af1f6032b873fa6df4d478/src/lib/epoll/iomgr_impl_epoll.cpp
    @ 00000000066ef1a8 _ZSt8__invokeIZN5iomgr18IOManagerEpollImpl19create_reactor_implERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEmjiOSt8functionIFvbEEEUlvE_JEENSt15__invoke_resultIT_JDpT0_EE4typeEOSG_DpOSH_
                       /usr/include/c++/11/bits/invoke.h:96
                       -> /var/home/centos/.conan/data/iomgr/11.2.2-33/oss/master/build/8f6113204e39cf5069af1f6032b873fa6df4d478/src/lib/epoll/iomgr_impl_epoll.cpp
    @ 00000000066ef155 _ZNSt6thread8_InvokerISt5tupleIJZN5iomgr18IOManagerEpollImpl19create_reactor_implERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEmjiOSt8functionIFvbEEEUlvE_EEE9_M_invokeIJLm0EEEEvSt12_Index_tupleIJXspT_EEE
                       /usr/include/c++/11/bits/std_thread.h:259
                       -> /var/home/centos/.conan/data/iomgr/11.2.2-33/oss/master/build/8f6113204e39cf5069af1f6032b873fa6df4d478/src/lib/epoll/iomgr_impl_epoll.cpp
    @ 00000000066ef129 _ZNSt6thread8_InvokerISt5tupleIJZN5iomgr18IOManagerEpollImpl19create_reactor_implERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEmjiOSt8functionIFvbEEEUlvE_EEEclEv
                       /usr/include/c++/11/bits/std_thread.h:266
                       -> /var/home/centos/.conan/data/iomgr/11.2.2-33/oss/master/build/8f6113204e39cf5069af1f6032b873fa6df4d478/src/lib/epoll/iomgr_impl_epoll.cpp
    @ 00000000066ef10d _ZNSt6thread11_State_implINS_8_InvokerISt5tupleIJZN5iomgr18IOManagerEpollImpl19create_reactor_implERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEmjiOSt8functionIFvbEEEUlvE_EEEEE6_M_runEv
                       /usr/include/c++/11/bits/std_thread.h:211
                       -> /var/home/centos/.conan/data/iomgr/11.2.2-33/oss/master/build/8f6113204e39cf5069af1f6032b873fa6df4d478/src/lib/epoll/iomgr_impl_epoll.cpp
```

Reply:
this error happens when trying to creating the first shard of a pg. I think this a problem caused by the issue here 
https://github.com/eBay/HomeObject/pull/159#discussion_r1545353794, I believe it is a bug.

let me try to fix it

Reply:
i have created a PR to solve this issue, PTAL
https://github.com/eBay/HomeStore/pull/380